### PR TITLE
Lack of base directory parameters to point at external drives if needed

### PR DESCRIPTION
**Pull request by r8d8**
Use default path and give the opportunity to change it.

Related to #15.

### CODE REVIEW

**Comment by dulanov** [comment]
@r8d8 Please give an issue reference not in PR title, but in description because of the issue link in title is not recognized by github.

**Comment by splix** [comment]
but why we need a separate path for keystore?

**Comment by dulanov** [comment]
Why not, for example keys can lie on external drive separately?

**Comment by r8d8** [comment]
Should we made provide only `--base-dir` argument and made keystore as nested dir inside it?

**Comment by splix** [comment]
@dulanov okaay. I just think it's a precocious feature. but ok

@r8d8 yes, by default it should be inside a basedir, like `basedir/%CHAIN%/keystore`

**Comment by splix** [comment]
let's use `default` as `%CHAIN%` for now. later it will be `testnet`, `mainnet`, etc

File: src/main.rs [hunk 1]
```diff
@@ -64,10 +67,20 @@ fn main() {
...
 
+    let base_path_str = args.flag_base_path
+        .parse::<String>()
+        .expect("Expect to parse base path");
+    let base_path;
+    if !base_path_str.is_empty() {
+        base_path = Some(Path::new(&base_path_str));
+    } else {
+        base_path = None;
+    }
```
Comment (dulanov) [inline src/main.rs]: `let base_path = if ...`

File: src/main.rs [hunk 2]
```diff
@@ -32,6 +33,8 @@ struct Args {
     flag_quiet: bool,
     flag_address: String,
     flag_client_address: String,
+    flag_base_path: String,
+    flag_key_path: String,
```
Comment (dulanov) [inline src/main.rs]: We decided not to use keypath, right?

File: src/storage.rs [hunk 3]
```diff
@@ -11,11 +11,12 @@ pub struct Storages<'a> {
...
-    pub fn new() -> Storages<'a> {
-        Storages {
-            //TODO use user home subdir
-            base_dir: Path::new("emerald_data"),
+    /// Create storage using user directory if specified,
+    /// or default path in other case.
+    pub fn new(path: Option<&'a Path>) -> Storages<'a> {
+        match path {
+            Some(p) => Storages { base_dir: p },
+            _ => Storages { base_dir: Path::new("emerald_data") },
```
Comment (dulanov) [inline src/storage.rs]: Why `emerald_data`, in issue stay:

Mac OS X: ~/Library/Emerald
Linux: ~/.emerald
Windows: %USERDIR%\.emerald

File: src/storage.rs [hunk 4]
```diff
@@ -68,3 +69,24 @@ impl<'a> ChainStorage<'a> {
...
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn should_use_default_path() {
+        let st = Storages::new(None);
+
+        assert_eq!(st.base_dir.as_os_str(),
+                   Path::new("emerald_data").as_os_str());
```
Comment (dulanov) [inline src/storage.rs]: You really need `.as_os_str()` here?

File: usage.txt [hunk 5]
```diff
@@ -12,3 +12,4 @@ Options:
     -q, --quiet                 No output printed to stdout.
     --address ADDR              Emerald wallet listen address. [default: 127.0.0.1:1920]
     --client-address ADDR       Ethereum client connect to address. [default: 127.0.0.1:8545]
+    --base-path PATH            Base directory path.
```
Comment (dulanov) [inline usage.txt]: Add information about default path.

**Comment by dulanov** [comment]
Please merge with upstream.

**Pull request merged**